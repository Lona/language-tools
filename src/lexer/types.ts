export type NextAction = { type: 'next'; value: string }

export type PushAction = { type: 'push'; value: string }

export type PopAction = { type: 'pop' }

export type Action = NextAction | PushAction | PopAction

export type Rule = {
  name: string
  pattern: string
  action?: Action
  discard: boolean
  // print: TokenPrintPattern
}

export type StateDefinition = {
  name: string
  rules: Rule[]
}

export type Position = {
  start: number
  end: number
}

export interface Token {
  type: string
  values: string[]
  position: Position
}

// export interface TokenWithPosition extends Token {
//   position: Position
// }

export type LexerDefinition = StateDefinition[]
