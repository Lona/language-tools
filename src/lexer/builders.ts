import {
  Action,
  Rule,
  Position,
  NextAction,
  PushAction,
  PopAction,
  StateDefinition,
  Token,
} from './types'

/* -------- Actions -------- */

export function nextAction(value: string): NextAction {
  return { type: 'next', value }
}

export function pushAction(value: string): PushAction {
  return { type: 'push', value }
}

export function pop(): PopAction {
  return { type: 'pop' }
}

/* -------- Rules -------- */

type RuleOptions = {
  discard?: boolean
  action?: Action
  // print?: TokenPrintPattern
}

/**
 * Match a keyword followed by a word boundary
 */
export function keyword(name: string): Rule {
  return rule(name, new RegExp(name + '\\b'))
}

export function rule(name: string): Rule
export function rule(name: string, pattern: string | RegExp): Rule
export function rule(
  name: string,
  pattern: string | RegExp,
  options: RuleOptions
): Rule
export function rule(
  name: string,
  pattern?: string | RegExp,
  options: RuleOptions = {}
): Rule {
  const source = pattern instanceof RegExp ? pattern.source : pattern

  return {
    name,
    pattern: source ?? name,
    discard: options.discard ?? false,
    // print: options.print ?? { type: 'literal', value: pattern },
    action: options.action,
  }
}

/* -------- Tokens -------- */

type TokenOptions = { value?: string; values?: string[]; position?: Position }

export function token(type: string): Token
export function token(type: string, value: string): Token
export function token(type: string, options: TokenOptions): Token

export function token(
  type: string,
  optionsOrValue?: TokenOptions | string
): Token {
  const options =
    typeof optionsOrValue === 'undefined'
      ? {}
      : typeof optionsOrValue === 'string'
      ? { values: [optionsOrValue] }
      : optionsOrValue

  return {
    type,
    values: options.values ?? [],
    position: options.position ?? { start: 0, end: 0 },
  }
}

/* -------- States -------- */

export type RuleShorthand =
  | string
  | [string]
  | [string, string | RegExp]
  | [string, string | RegExp, string]
  | [string, string | RegExp, RuleOptions]

/**
 * Support Monarch lexer shorthand patterns
 */
export function normalizeRule(rule: Rule | RuleShorthand): Rule {
  let arrayOrRule = typeof rule === 'string' ? [rule] : rule

  if (arrayOrRule instanceof Array) {
    const [name, pattern, options] = arrayOrRule

    const source = pattern instanceof RegExp ? pattern.source : pattern

    return {
      name,
      pattern: source ?? name,
      discard: false,
      ...(typeof options === 'string'
        ? { action: { type: 'next', value: options } }
        : options),
    }
  }

  return arrayOrRule
}

export function state(
  name: string,
  rules: Parameters<typeof normalizeRule>[0][]
): StateDefinition {
  if (rules.length === 0) {
    throw new Error('A lexer state must contain at least 1 rule')
  }

  return { name, rules: rules.map(normalizeRule) }
}
