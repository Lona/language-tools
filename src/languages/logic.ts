import { state, token, keyword } from '../lexer/builders'
import { Lexer } from '../lexer/Lexer'
import { LexerDefinition, Token } from '../lexer/types'
import {
  consume,
  language,
  many,
  option,
  printIf,
  printIndentedMany,
  printLine,
  printLiteral,
  sequence,
  thunk,
  typedOr,
} from '../parser/builders'
import { Parser, ParserPattern } from '../parser/Parser'
import { Printer } from '../printer/Printer'
import { Shaper } from '../shaper/Shaper'
import { createTemplates } from '../templates/templates'
import countGroups from '../utils/countGroups'

export const lexerDefinition: LexerDefinition = [
  state('main', [
    keyword('true'),
    keyword('false'),
    keyword('enum'),
    keyword('case'),
    keyword('struct'),
    keyword('extension'),
    keyword('let'),
    keyword('func'),
    keyword('return'),
    keyword('import'),

    // Special chars
    '#',
    '@',
    ['[', /\[/],
    [']', /\]/],
    ['{', /\{/],
    ['}', /\}/],
    ['(', /\(/],
    [')', /\)/],
    ',',
    ':',
    '->',
    '=',
    '>',
    '<',
    ['.', /\./],

    ['identifier', /([_a-zA-Z][_a-zA-Z0-9]*)/],
    ['string', /"([^"]*)"/],
    ['float', /((?:\+|-)?[0-9]+\.[0-9]+)/],
    ['integer', /((?:\+|-)?[0-9]+)/],
    ['comment', /\/\*([\S\s]*?)\*\//],
    ['whitespace', /[ \t\n\r]+/, { discard: true }],
  ]),
]

const blockDeclaration = (kind: string, body: ParserPattern): ParserPattern =>
  sequence([
    option(() => patterns.comment, 'comment'),
    () => patterns.attributeList,
    consume(kind),
    printLine(),
    thunk(() => patterns.identifier, 'name'),
    option(() => patterns.genericIdentifierList, {
      as: 'generics',
      select: 'generics',
      defaultValue: [],
    }),
    printLine(),
    consume('{'),
    printIndentedMany(
      many(body, {
        as: 'body',
        separator: printLine('hard'),
      }),
      { padding: 'hard' }
    ),
    consume('}'),
  ])

export const patterns = language({
  program: many(() => patterns.declaration, {
    as: 'declarations',
    separator: printLine('hard'),
  }),
  declaration: typedOr(
    [
      ['variableDeclaration', () => patterns.variableDeclaration],
      ['structDeclaration', () => patterns.structDeclaration],
      ['namespaceDeclaration', () => patterns.namespaceDeclaration],
      ['functionDeclaration', () => patterns.functionDeclaration],
      ['enumDeclaration', () => patterns.enumDeclaration],
      ['importDeclaration', () => patterns.importDeclaration],
    ],
    {
      as: '.',
    }
  ),
  importDeclaration: [
    consume('import'),
    printLine(),
    thunk(() => patterns.identifier, 'name'),
  ],
  comment: [
    printLiteral('/*'),
    consume('comment', 'comment'),
    printLiteral('*/'),
    printLine('hard'),
  ],
  attributeList: printIf(
    many(() => patterns.attribute, {
      as: 'attributes',
      separator: printLine('hard'),
    }),
    { reference: 'attributes', after: printLine('hard') }
  ),
  genericIdentifierList: [
    consume('<'),
    many(() => patterns.identifier, {
      separator: [consume(','), printLine()],
      as: 'generics',
    }),
    consume('>'),
  ],
  variableDeclaration: [
    option(() => patterns.comment, 'comment'),
    () => patterns.attributeList,
    consume('let'),
    printLine(),
    thunk(() => patterns.identifier, 'name'),
    consume(':'),
    printLine(),
    thunk(() => patterns.typeAnnotation, 'annotation'),
    printLine(),
    consume('='),
    printLine(),
    thunk(() => patterns.expression, 'initializer'),
  ],
  functionParameter: [
    thunk(() => patterns.identifier, 'name'),
    consume(':'),
    printLine(),
    thunk(() => patterns.typeAnnotation, 'annotation'),
  ],
  functionDeclaration: [
    option(() => patterns.comment, 'comment'),
    () => patterns.attributeList,
    consume('func'),
    printLine(),
    thunk(() => patterns.identifier, 'name'),
    option(() => patterns.genericIdentifierList, {
      as: 'generics',
      select: 'generics',
      defaultValue: [],
    }),
    consume('('),
    printIndentedMany(
      many(() => patterns.functionParameter, {
        as: 'arguments',
        separator: [consume(','), printLine()],
      }),
      { padding: 'soft' }
    ),
    consume(')'),
    printLine(),
    consume('->'),
    printLine(),
    thunk(() => patterns.typeAnnotation, 'returnType'),
    printLine(),
    consume('{'),
    consume('}'),
  ],
  structDeclaration: blockDeclaration(
    'struct',
    thunk(() => patterns.declaration)
  ),
  namespaceDeclaration: blockDeclaration(
    'extension',
    thunk(() => patterns.declaration)
  ),
  enumDeclaration: blockDeclaration(
    'enum',
    thunk(() => patterns.enumCase)
  ),
  enumCase: [
    option(() => patterns.comment, 'comment'),
    consume('case'),
    printLine(),
    thunk(() => patterns.identifier, 'name'),
    consume('('),
    many(() => patterns.associatedValue, {
      as: 'associatedValues',
      separator: [consume(','), printLine()],
    }),
    consume(')'),
  ],
  expression: typedOr(
    [
      ['functionCallExpression', () => patterns.functionCallExpression],
      ['memberExpression', () => patterns.memberExpression],
      ['identifierExpression', () => patterns.identifierExpression],
      ['literalExpression', () => patterns.literalExpression],
    ],
    {
      parallel: true,
      as: '.',
    }
  ),
  functionCallExpression: [
    thunk(() => patterns.expression, 'expression'),
    consume('('),
    printIndentedMany(
      many(() => patterns.functionCallArgument, {
        as: 'arguments',
        separator: [consume(','), printLine()],
      }),
      { padding: 'soft' }
    ),
    consume(')'),
  ],
  functionCallArgument: [
    option(
      [
        sequence([() => patterns.identifier, consume(':'), printLine()], {
          as: 'identifier',
        }),
      ],
      {
        as: 'label',
        select: 'identifier',
      }
    ),
    thunk(() => patterns.expression, 'expression'),
  ],
  associatedValue: [
    option(
      sequence([() => patterns.identifier, consume(':'), printLine()], {
        as: 'identifier',
      }),
      {
        as: 'label',
        select: 'identifier',
      }
    ),
    thunk(() => patterns.typeAnnotation, 'annotation'),
  ],
  memberExpression: [
    thunk(() => patterns.expression, 'expression'),
    consume('.'),
    thunk(() => patterns.expression, 'member'),
  ],
  identifierExpression: [() => patterns.identifier],
  literalExpression: typedOr(
    [
      ['boolean', () => patterns.booleanLiteral],
      ['number', () => patterns.numberLiteral],
      ['string', () => patterns.stringLiteral],
      ['array', () => patterns.arrayLiteral],
    ],
    {
      as: '.',
    }
  ),
  booleanLiteral: typedOr(
    [
      ['true', consume('true')],
      ['false', consume('false')],
    ],
    {
      as: '.',
    }
  ),
  numberLiteral: typedOr(
    [
      ['integer', consume('integer', '.')],
      ['float', consume('float', '.')],
    ],
    {
      as: '.',
    }
  ),
  stringLiteral: [
    printLiteral('"'),
    consume('string', 'value'),
    printLiteral('"'),
  ],
  arrayLiteral: [
    consume('['),
    printIndentedMany(
      many(() => patterns.expression, {
        as: 'body',
        separator: sequence([consume(','), printLine()]),
      }),
      { padding: 'soft' }
    ),
    consume(']'),
  ],
  typeAnnotation: [
    () => patterns.identifier,
    option(
      [
        consume('<'),
        many(() => patterns.typeAnnotation, {
          as: 'genericArguments',
          separator: [consume(','), printLine()],
        }),
        consume('>'),
      ],
      {
        as: 'genericArguments',
        select: 'genericArguments',
        defaultValue: [],
      }
    ),
  ],
  identifier: consume('identifier', 'value'),
  attribute: [consume('@'), consume('identifier', 'value')],
})

export const lexer = new Lexer(lexerDefinition)

export const parser = new Parser({
  matchToken: (token: Token, type: string) => token.type === type,
  tokenToValue: (token: Token) => token,
})

const tokenToString = (token: Token): string =>
  token.values.length > 0 ? token.values[0] : token.type

export const printer = new Printer({
  tokenToString,
})

const tokenTypesWithValues: string[] = lexerDefinition
  .flatMap((definition) => definition.rules)
  .filter((rule) => countGroups(rule.pattern) > 0)
  .map((rule) => rule.name)

export const shaper = new Shaper<Token, string, any>({
  shapeToToken: (shape: string, type: string) =>
    tokenTypesWithValues.includes(type)
      ? token(type, { values: [shape] })
      : token(type),
  tokenToShape: tokenToString,
  stringToShape: (string: string, type: string): string =>
    tokenTypesWithValues.includes(type) ? string : type,
  orShapeToValue: (shape) => shape,
  orValueToShape: (value) => value,
})

export const templates = createTemplates({
  lexer,
  parser,
  shaper,
  patterns,
  printer,
})
