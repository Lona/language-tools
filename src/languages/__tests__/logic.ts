import fs from 'fs'
import path from 'path'
import { resetId } from '../../parser/builders'
import { Pattern } from '../../parser/types'
import { lexer, parser, patterns, printer, shaper } from '../logic'
import { inspect } from 'util'

function expectParse(
  source: string,
  pattern: Pattern<string>,
  expectedShape?: unknown
) {
  const tokens = lexer.tokenize(source, undefined, tokenizeOptions)
  const result = parser.parseOne(pattern, tokens)

  expect(printer.print(result)).toEqual(source)

  const shape = shaper.extract(result)

  if (expectedShape) {
    expect(shape).toEqual(expectedShape)
  } else {
    expect(shape).toMatchSnapshot()
  }

  const recreated = shaper.inject(shape, pattern)
  expect(recreated).toEqual(result)
}

const tokenizeOptions = {
  positionTracking: false,
}

describe('Logic Language', () => {
  beforeEach(() => {
    resetId()
  })

  it('tokenizes', () => {
    const tokens = lexer.tokenize('let x = 123')
    expect(tokens.length).toEqual(4)
  })

  it('parses identifier', () => {
    expectParse('foo', patterns.identifier, { value: 'foo' })
  })

  it('parses typeAnnotation', () => {
    expectParse('Foo', patterns.typeAnnotation, {
      value: 'Foo',
      genericArguments: [],
    })
    expectParse('Foo<A>', patterns.typeAnnotation, {
      value: 'Foo',
      genericArguments: [{ value: 'A', genericArguments: [] }],
    })
    expectParse('Foo<A, B>', patterns.typeAnnotation, {
      value: 'Foo',
      genericArguments: [
        { value: 'A', genericArguments: [] },
        { value: 'B', genericArguments: [] },
      ],
    })
  })

  it('parses program', () => {
    expectParse(`let foo: Number = x\nstruct Foo {}`, patterns.program)
  })

  it('parses declaration', () => {
    expectParse('let foo: Number = x', patterns.declaration)
    expectParse('struct Foo {}', patterns.declaration)
    expectParse('enum Foo {}', patterns.declaration)
  })

  it('parses import declaration', () => {
    expectParse('import Foo', patterns.importDeclaration)
  })

  it('parses function declaration', () => {
    expectParse('func foo() -> Number {}', patterns.functionDeclaration)
    expectParse(
      'func foo(a: b, c: d) -> Number {}',
      patterns.functionDeclaration
    )
    expectParse('func foo<A>() -> Number {}', patterns.functionDeclaration)
    expectParse('func foo<A, B>() -> Number {}', patterns.functionDeclaration)
  })

  it('parses variable declaration', () => {
    expectParse('let foo: Number = x', patterns.variableDeclaration)
    expectParse('let foo: Number = 1', patterns.variableDeclaration)
    expectParse('@attr\nlet foo: Number = x', patterns.variableDeclaration)
    expectParse(
      '@attr1\n@attr2\nlet foo: Number = x',
      patterns.variableDeclaration
    )
    expectParse(
      '/* comment */\nlet foo: Number = x',
      patterns.variableDeclaration
    )
  })

  it('parses structs', () => {
    expectParse('struct Foo {}', patterns.structDeclaration)
    expectParse(
      `struct Foo {
  let foo: Number = x
}`,
      patterns.structDeclaration
    )
    expectParse(
      `struct Foo {
  let foo: Number = x
  let bar: String = y
}`,
      patterns.structDeclaration
    )
    expectParse('struct Foo<A> {}', patterns.structDeclaration)
  })

  it('parses enums', () => {
    expectParse('enum Foo {}', patterns.enumDeclaration)
    expectParse(
      `enum Foo {
  case a()
}`,
      patterns.enumDeclaration
    )
    expectParse(
      `enum Foo {
  case a(param: Number)
}`,
      patterns.enumDeclaration
    )
    expectParse(
      `enum Foo {
  case a(Number)
}`,
      patterns.enumDeclaration
    )
  })

  it('parses boolean literal', () => {
    expectParse('true', patterns.booleanLiteral)
    expectParse('false', patterns.booleanLiteral)
  })

  it('parses number literal', () => {
    expectParse('1.0', patterns.numberLiteral)
    expectParse('1', patterns.numberLiteral)
  })

  it('parses string literal', () => {
    expectParse('"abc"', patterns.stringLiteral)
    expectParse('""', patterns.stringLiteral)
  })

  it('parses array literal', () => {
    expectParse('[]', patterns.arrayLiteral)
    expectParse('[1, 2, 3]', patterns.arrayLiteral)
  })

  it('parses literal expression', () => {
    expectParse('1', patterns.literalExpression)
    expectParse('true', patterns.literalExpression)
  })

  it('parses identifier expression', () => {
    expectParse('x', patterns.identifierExpression)
  })

  it('parses function calls', () => {
    expectParse('foo()', patterns.functionCallExpression)
    expectParse('foo(a)', patterns.functionCallExpression)
    expectParse('foo(a, b)', patterns.functionCallExpression)
    expectParse('foo(a: b)', patterns.functionCallExpression)
    expectParse('foo(a: b, c: d)', patterns.functionCallExpression)
  })

  it('parses function call argument', () => {
    expectParse('a: b', patterns.functionCallArgument)
    expectParse('a', patterns.functionCallArgument)
  })

  it('parses member expressions', () => {
    expectParse('foo.bar', patterns.memberExpression)
    expectParse('foo.bar()', patterns.memberExpression)
  })

  it('parses expressions', () => {
    expectParse('x', patterns.expression)
    expectParse('1', patterns.expression)
    expectParse('[1, 2]', patterns.expression)
    expectParse('foo()', patterns.expression)
    expectParse('foo.bar', patterns.expression)
  })

  it('parses Prelude.logic', () => {
    const read = (name: string) =>
      fs.readFileSync(path.join(__dirname, 'mocks', name), 'utf8')

    const tokens = lexer.tokenize(
      read('Prelude.logic'),
      undefined,
      tokenizeOptions
    )

    const result = parser.parseOne(patterns.program, tokens)

    shaper.extract(result)
  })

  it('parses TextStyle.logic', () => {
    const read = (name: string) =>
      fs.readFileSync(path.join(__dirname, 'mocks', name), 'utf8')

    const tokens = lexer.tokenize(
      read('TextStyle.logic'),
      undefined,
      tokenizeOptions
    )

    try {
      const result = parser.parseOne(patterns.program, tokens)
      shaper.extract(result)
    } catch (e) {
      console.log(inspect(e, false, null))
      expect(false).toBe(true)
    }
  })
})
