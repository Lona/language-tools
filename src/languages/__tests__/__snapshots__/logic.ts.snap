// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logic Language parses array literal 1`] = `
Object {
  "body": Array [],
}
`;

exports[`Logic Language parses array literal 2`] = `
Object {
  "body": Array [
    Object {
      "type": "literalExpression",
      "value": Object {
        "type": "number",
        "value": Object {
          "type": "integer",
          "value": "1",
        },
      },
    },
    Object {
      "type": "literalExpression",
      "value": Object {
        "type": "number",
        "value": Object {
          "type": "integer",
          "value": "2",
        },
      },
    },
    Object {
      "type": "literalExpression",
      "value": Object {
        "type": "number",
        "value": Object {
          "type": "integer",
          "value": "3",
        },
      },
    },
  ],
}
`;

exports[`Logic Language parses boolean literal 1`] = `
Object {
  "type": "true",
  "value": null,
}
`;

exports[`Logic Language parses boolean literal 2`] = `
Object {
  "type": "false",
  "value": null,
}
`;

exports[`Logic Language parses declaration 1`] = `
Object {
  "type": "variableDeclaration",
  "value": Object {
    "annotation": Object {
      "genericArguments": Array [],
      "value": "Number",
    },
    "attributes": Array [],
    "initializer": Object {
      "type": "identifierExpression",
      "value": Object {
        "value": "x",
      },
    },
    "name": Object {
      "value": "foo",
    },
  },
}
`;

exports[`Logic Language parses declaration 2`] = `
Object {
  "type": "structDeclaration",
  "value": Object {
    "attributes": Array [],
    "body": Array [],
    "generics": Array [],
    "name": Object {
      "value": "Foo",
    },
  },
}
`;

exports[`Logic Language parses declaration 3`] = `
Object {
  "type": "enumDeclaration",
  "value": Object {
    "attributes": Array [],
    "body": Array [],
    "generics": Array [],
    "name": Object {
      "value": "Foo",
    },
  },
}
`;

exports[`Logic Language parses enums 1`] = `
Object {
  "attributes": Array [],
  "body": Array [],
  "generics": Array [],
  "name": Object {
    "value": "Foo",
  },
}
`;

exports[`Logic Language parses enums 2`] = `
Object {
  "attributes": Array [],
  "body": Array [
    Object {
      "associatedValues": Array [],
      "name": Object {
        "value": "a",
      },
    },
  ],
  "generics": Array [],
  "name": Object {
    "value": "Foo",
  },
}
`;

exports[`Logic Language parses enums 3`] = `
Object {
  "attributes": Array [],
  "body": Array [
    Object {
      "associatedValues": Array [
        Object {
          "annotation": Object {
            "genericArguments": Array [],
            "value": "Number",
          },
          "label": Object {
            "value": "param",
          },
        },
      ],
      "name": Object {
        "value": "a",
      },
    },
  ],
  "generics": Array [],
  "name": Object {
    "value": "Foo",
  },
}
`;

exports[`Logic Language parses enums 4`] = `
Object {
  "attributes": Array [],
  "body": Array [
    Object {
      "associatedValues": Array [
        Object {
          "annotation": Object {
            "genericArguments": Array [],
            "value": "Number",
          },
        },
      ],
      "name": Object {
        "value": "a",
      },
    },
  ],
  "generics": Array [],
  "name": Object {
    "value": "Foo",
  },
}
`;

exports[`Logic Language parses expressions 1`] = `
Object {
  "type": "identifierExpression",
  "value": Object {
    "value": "x",
  },
}
`;

exports[`Logic Language parses expressions 2`] = `
Object {
  "type": "literalExpression",
  "value": Object {
    "type": "number",
    "value": Object {
      "type": "integer",
      "value": "1",
    },
  },
}
`;

exports[`Logic Language parses expressions 3`] = `
Object {
  "type": "literalExpression",
  "value": Object {
    "type": "array",
    "value": Object {
      "body": Array [
        Object {
          "type": "literalExpression",
          "value": Object {
            "type": "number",
            "value": Object {
              "type": "integer",
              "value": "1",
            },
          },
        },
        Object {
          "type": "literalExpression",
          "value": Object {
            "type": "number",
            "value": Object {
              "type": "integer",
              "value": "2",
            },
          },
        },
      ],
    },
  },
}
`;

exports[`Logic Language parses expressions 4`] = `
Object {
  "type": "functionCallExpression",
  "value": Object {
    "arguments": Array [],
    "expression": Object {
      "type": "identifierExpression",
      "value": Object {
        "value": "foo",
      },
    },
  },
}
`;

exports[`Logic Language parses expressions 5`] = `
Object {
  "type": "memberExpression",
  "value": Object {
    "expression": Object {
      "type": "identifierExpression",
      "value": Object {
        "value": "foo",
      },
    },
    "member": Object {
      "type": "identifierExpression",
      "value": Object {
        "value": "bar",
      },
    },
  },
}
`;

exports[`Logic Language parses function call argument 1`] = `
Object {
  "expression": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "b",
    },
  },
  "label": Object {
    "value": "a",
  },
}
`;

exports[`Logic Language parses function call argument 2`] = `
Object {
  "expression": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "a",
    },
  },
}
`;

exports[`Logic Language parses function calls 1`] = `
Object {
  "arguments": Array [],
  "expression": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "foo",
    },
  },
}
`;

exports[`Logic Language parses function calls 2`] = `
Object {
  "arguments": Array [
    Object {
      "expression": Object {
        "type": "identifierExpression",
        "value": Object {
          "value": "a",
        },
      },
    },
  ],
  "expression": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "foo",
    },
  },
}
`;

exports[`Logic Language parses function calls 3`] = `
Object {
  "arguments": Array [
    Object {
      "expression": Object {
        "type": "identifierExpression",
        "value": Object {
          "value": "a",
        },
      },
    },
    Object {
      "expression": Object {
        "type": "identifierExpression",
        "value": Object {
          "value": "b",
        },
      },
    },
  ],
  "expression": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "foo",
    },
  },
}
`;

exports[`Logic Language parses function calls 4`] = `
Object {
  "arguments": Array [
    Object {
      "expression": Object {
        "type": "identifierExpression",
        "value": Object {
          "value": "b",
        },
      },
      "label": Object {
        "value": "a",
      },
    },
  ],
  "expression": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "foo",
    },
  },
}
`;

exports[`Logic Language parses function calls 5`] = `
Object {
  "arguments": Array [
    Object {
      "expression": Object {
        "type": "identifierExpression",
        "value": Object {
          "value": "b",
        },
      },
      "label": Object {
        "value": "a",
      },
    },
    Object {
      "expression": Object {
        "type": "identifierExpression",
        "value": Object {
          "value": "d",
        },
      },
      "label": Object {
        "value": "c",
      },
    },
  ],
  "expression": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "foo",
    },
  },
}
`;

exports[`Logic Language parses function declaration 1`] = `
Object {
  "arguments": Array [],
  "attributes": Array [],
  "generics": Array [],
  "name": Object {
    "value": "foo",
  },
  "returnType": Object {
    "genericArguments": Array [],
    "value": "Number",
  },
}
`;

exports[`Logic Language parses function declaration 2`] = `
Object {
  "arguments": Array [
    Object {
      "annotation": Object {
        "genericArguments": Array [],
        "value": "b",
      },
      "name": Object {
        "value": "a",
      },
    },
    Object {
      "annotation": Object {
        "genericArguments": Array [],
        "value": "d",
      },
      "name": Object {
        "value": "c",
      },
    },
  ],
  "attributes": Array [],
  "generics": Array [],
  "name": Object {
    "value": "foo",
  },
  "returnType": Object {
    "genericArguments": Array [],
    "value": "Number",
  },
}
`;

exports[`Logic Language parses function declaration 3`] = `
Object {
  "arguments": Array [],
  "attributes": Array [],
  "generics": Array [
    Object {
      "value": "A",
    },
  ],
  "name": Object {
    "value": "foo",
  },
  "returnType": Object {
    "genericArguments": Array [],
    "value": "Number",
  },
}
`;

exports[`Logic Language parses function declaration 4`] = `
Object {
  "arguments": Array [],
  "attributes": Array [],
  "generics": Array [
    Object {
      "value": "A",
    },
    Object {
      "value": "B",
    },
  ],
  "name": Object {
    "value": "foo",
  },
  "returnType": Object {
    "genericArguments": Array [],
    "value": "Number",
  },
}
`;

exports[`Logic Language parses identifier expression 1`] = `
Object {
  "value": "x",
}
`;

exports[`Logic Language parses import declaration 1`] = `
Object {
  "name": Object {
    "value": "Foo",
  },
}
`;

exports[`Logic Language parses literal expression 1`] = `
Object {
  "type": "number",
  "value": Object {
    "type": "integer",
    "value": "1",
  },
}
`;

exports[`Logic Language parses literal expression 2`] = `
Object {
  "type": "boolean",
  "value": Object {
    "type": "true",
    "value": null,
  },
}
`;

exports[`Logic Language parses member expressions 1`] = `
Object {
  "expression": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "foo",
    },
  },
  "member": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "bar",
    },
  },
}
`;

exports[`Logic Language parses member expressions 2`] = `
Object {
  "expression": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "foo",
    },
  },
  "member": Object {
    "type": "functionCallExpression",
    "value": Object {
      "arguments": Array [],
      "expression": Object {
        "type": "identifierExpression",
        "value": Object {
          "value": "bar",
        },
      },
    },
  },
}
`;

exports[`Logic Language parses number literal 1`] = `
Object {
  "type": "float",
  "value": "1.0",
}
`;

exports[`Logic Language parses number literal 2`] = `
Object {
  "type": "integer",
  "value": "1",
}
`;

exports[`Logic Language parses program 1`] = `
Object {
  "declarations": Array [
    Object {
      "type": "variableDeclaration",
      "value": Object {
        "annotation": Object {
          "genericArguments": Array [],
          "value": "Number",
        },
        "attributes": Array [],
        "initializer": Object {
          "type": "identifierExpression",
          "value": Object {
            "value": "x",
          },
        },
        "name": Object {
          "value": "foo",
        },
      },
    },
    Object {
      "type": "structDeclaration",
      "value": Object {
        "attributes": Array [],
        "body": Array [],
        "generics": Array [],
        "name": Object {
          "value": "Foo",
        },
      },
    },
  ],
}
`;

exports[`Logic Language parses string literal 1`] = `
Object {
  "value": "abc",
}
`;

exports[`Logic Language parses string literal 2`] = `
Object {
  "value": "",
}
`;

exports[`Logic Language parses structs 1`] = `
Object {
  "attributes": Array [],
  "body": Array [],
  "generics": Array [],
  "name": Object {
    "value": "Foo",
  },
}
`;

exports[`Logic Language parses structs 2`] = `
Object {
  "attributes": Array [],
  "body": Array [
    Object {
      "type": "variableDeclaration",
      "value": Object {
        "annotation": Object {
          "genericArguments": Array [],
          "value": "Number",
        },
        "attributes": Array [],
        "initializer": Object {
          "type": "identifierExpression",
          "value": Object {
            "value": "x",
          },
        },
        "name": Object {
          "value": "foo",
        },
      },
    },
  ],
  "generics": Array [],
  "name": Object {
    "value": "Foo",
  },
}
`;

exports[`Logic Language parses structs 3`] = `
Object {
  "attributes": Array [],
  "body": Array [
    Object {
      "type": "variableDeclaration",
      "value": Object {
        "annotation": Object {
          "genericArguments": Array [],
          "value": "Number",
        },
        "attributes": Array [],
        "initializer": Object {
          "type": "identifierExpression",
          "value": Object {
            "value": "x",
          },
        },
        "name": Object {
          "value": "foo",
        },
      },
    },
    Object {
      "type": "variableDeclaration",
      "value": Object {
        "annotation": Object {
          "genericArguments": Array [],
          "value": "String",
        },
        "attributes": Array [],
        "initializer": Object {
          "type": "identifierExpression",
          "value": Object {
            "value": "y",
          },
        },
        "name": Object {
          "value": "bar",
        },
      },
    },
  ],
  "generics": Array [],
  "name": Object {
    "value": "Foo",
  },
}
`;

exports[`Logic Language parses structs 4`] = `
Object {
  "attributes": Array [],
  "body": Array [],
  "generics": Array [
    Object {
      "value": "A",
    },
  ],
  "name": Object {
    "value": "Foo",
  },
}
`;

exports[`Logic Language parses variable declaration 1`] = `
Object {
  "annotation": Object {
    "genericArguments": Array [],
    "value": "Number",
  },
  "attributes": Array [],
  "initializer": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "x",
    },
  },
  "name": Object {
    "value": "foo",
  },
}
`;

exports[`Logic Language parses variable declaration 2`] = `
Object {
  "annotation": Object {
    "genericArguments": Array [],
    "value": "Number",
  },
  "attributes": Array [],
  "initializer": Object {
    "type": "literalExpression",
    "value": Object {
      "type": "number",
      "value": Object {
        "type": "integer",
        "value": "1",
      },
    },
  },
  "name": Object {
    "value": "foo",
  },
}
`;

exports[`Logic Language parses variable declaration 3`] = `
Object {
  "annotation": Object {
    "genericArguments": Array [],
    "value": "Number",
  },
  "attributes": Array [
    Object {
      "value": "attr",
    },
  ],
  "initializer": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "x",
    },
  },
  "name": Object {
    "value": "foo",
  },
}
`;

exports[`Logic Language parses variable declaration 4`] = `
Object {
  "annotation": Object {
    "genericArguments": Array [],
    "value": "Number",
  },
  "attributes": Array [
    Object {
      "value": "attr1",
    },
    Object {
      "value": "attr2",
    },
  ],
  "initializer": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "x",
    },
  },
  "name": Object {
    "value": "foo",
  },
}
`;

exports[`Logic Language parses variable declaration 5`] = `
Object {
  "annotation": Object {
    "genericArguments": Array [],
    "value": "Number",
  },
  "attributes": Array [],
  "comment": Object {
    "comment": " comment ",
  },
  "initializer": Object {
    "type": "identifierExpression",
    "value": Object {
      "value": "x",
    },
  },
  "name": Object {
    "value": "foo",
  },
}
`;
